### ISSUE TITLE
How do you update AutoLayout constraints in IGListCollectionView? What about Nested IGListCollectionView's?

### ISSUE THREAD

**Issue by egarlock** [Instagram/IGListKit#332]
Hello All,

I am currently working on a project in which I need to create multiple levels of expanding views (see screenshot below).

<img src="https://cloud.githubusercontent.com/assets/2692482/21201362/4e4b0902-c210-11e6-91d9-e4fd606811ba.PNG" width="200"><img src="https://cloud.githubusercontent.com/assets/2692482/21201370/53f34d88-c210-11e6-9223-428f43381006.PNG" width="200"><img src="https://cloud.githubusercontent.com/assets/2692482/21201375/57049a04-c210-11e6-8a17-37d84df7b292.PNG" width="200">

I have this implemented with IGListKit following something similar to the 'NestedAdapterViewController' View Controller in the provided example project. I have a SectionSectionCollection which is returned for my objects that are sections and creates a single cell with a UICollectionView within it. The SectionSectionCollection has an adapter and is the IGListAdapterDataSource for the cell's UICollectionView. As the datasource if this section has a section within it's IGListDiffable it will return an additional SectionSectionController and thus we have nested IGListCollectionView's! Wooo!

**Where I am having a problem is the fact that the 'SectionCells' have constraints that will get updated to expand or collapse. I am not sure exactly how to go about having the IGListCollectionView's update to be the height of their content as child views are expanded.**

The solution I have now works for the first level that is 'Section 1' and the second level 'Section 2'. I am setting the cell's button to a target selector within the 'SectionSectionController' and calling `[self.collectionContext reloadSectionController:self];` on the parent SectionSectionController.

## Q: How do I put this into practice with nested IGListCollectionView's, is the way I am currently doing it along the right track?

<pre>
<b>#pragma mark IBActions</b>
- (void)cellButton_TouchUpInside:(id)sender {
    _isExpanded = !self.isExpanded;
    
    [self reloadSectionController];
}

...

<b>#pragma mark &lt;IGListSectionType&gt;</b>
- (UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index {
    ExpandingCell *cell = (ExpandingCell *)[self.collectionContext dequeueReusableCellOfClass:[ExpandingCell class] forSectionController:self atIndex:index];

    [cell setIsExpanded:self.isExpanded];
    [cell setSection:self.object];
    [cell.cellButton addTarget:self action:@selector(cellButton_TouchUpInside:) forControlEvents:UIControlEventTouchUpInside];
    [self.adapter setCollectionView:cell.collectionView];
    
    return cell;
}

...

<b>#pragma mark Private API</b>
- (void)reloadSectionController {
    if(!!self.parentSectionController) {
        [self.parentSectionController reloadSectionController];
    } else {
        [self.collectionContext reloadSectionController:self];
    }
}
</pre>

## New issue checklist

- [ **YES** ] I have reviewed the [`README`](https://github.com/Instagram/IGListKit/blob/master/README.md) and [documentation](http://instagram.github.io/IGListKit)
- [ **YES** ] I have searched [existing issues](https://github.com/Instagram/IGListKit/issues) and this is not a duplicate

### General information

- `IGListKit` version: **2.0.0**
- iOS version(s): **9.0**
- CocoaPods/Carthage version: **CocoaPods**
- Xcode version: **8.1**
- Devices/Simulators affected: **N/A**
- Reproducible in the demo project? (Yes/No): **N/A**
- Related issues: **N/A**

**Comment by rnystrom** [comment]
Oh man, this is tricky 🤔 . I think I understand your setup:

- Vertical collection view + adapter
- Parent section controller that returns a single `SectionCell`
- `SectionCell` has an embedded vertical collection view + adapter
- `SectionCell` height should be the content size of the embedded collection view

Right? YMMV, I've never done this before.

One idea worth trying is the parent section controller owning the `UICollectionViewLayout` that powers the embedded collection view. Attach the layout when creating your cell:

```objc
@interface ParentSectionController
@property IGListAdapter *adapter;
@property UICollectionViewFlowLayout *layout;
@end

- (CGSize)sizeForItemAtIndex:(NSInteger)index {
  return CGSizeMake(self.collectionContext.containerSize.width, 
    self.layout. collectionViewContentSize.height);
}

- (UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index {
  SectionCell *cell = [self.collectionContext deque...];
  cell.collectionView.collectionViewLayout = self.layout;
  self.adapter.collectionView = cell.collectionView;
  return cell;
}
```

If that doesn't work, the other option is having the parent section controller own the embedded collection view itself and just `addSubview:` when configuring the cell. This is probably less performant, but might work.

cc @jessesquires for other ideas?

**Comment by egarlock** [comment]
@rnystrom  That is pretty much my setup.

Except:
- the adapter for the cell's UICollecitonView is owned by the `SectionSectionController` that creates it.

Also right now the `SectionCell` creates a basic `UICollectionViewFlowLayout` when it creates the `IGListCollectionView`. I'll give your suggestions a try and let you know how they work out!

**Comment by rnystrom** [comment]
Oh wow so there are multiple layers of nesting, dang! That's pretty cool, but tough. Let me know how playing w/ who owns the layout/collectionView goes.

FWIW we do the same thing, having our cells create the layout+collectionView, but we don't really have any requirements quite like this.

**Comment by egarlock** [comment]
I was unable to get this to work with the UICollectionViewFlowLayout owned by the SectionController. I decided to try to figure this out with a more simple problem, i.e. without expanding / collapsing the `SectionCell`s.

I can add all the items and adjust the SectionCell's `constraintCollectionViewHeight.constant` on `preferredLayoutAttributesFittingAttributes:`. This will set the correct height of collection view. The collection view however will not update unless I start to scroll a bit.

What am I missing here?

**Comment by rnystrom** [comment]
@egarlock this is sort of a "chicken and egg" problem, since the parent will have to size itself **first**, then size the child. However the child is driving the size of the parent.

What about after the initial sizing of the child calling `[self.collectionContext reloadSectionController:self]`?

**Comment by egarlock** [comment]
[Demo](https://github.com/egarlock/IGListKitDemo)

I went ahead and made a repo to show you what I'm currently doing. It's hard to explain. Everything works for the first 2 levels and I'm using `[self.collectionContext reloadSectionController:self]`.

**Comment by egarlock** [comment]
I decided to go with custom expandable views for now. Thanks for the help @rnystrom!

**Comment by rnystrom** [comment]
Sure thing! Closing this as it sounds like you're unblocked. If you come up w/ any interesting solutions for this, more details or a blog post or something would be really helpful for others to learn from. Pretty interesting problem!

**Comment by ZacharyKhan** [comment]
Can I bring this thread back to life and ask a related question? 

I have nested section controllers that go 3 levels deep. It's very similar to Instagram's Profile UI. I'll break it down to make sure I'm being accurate here - let's say Instagram's Profile UI was broken down as such from top to bottom:

- TopSectionController (_FollowersCell, UsernameCell, BioCell, ButtonsCell_)
- StoriesSectionController (_NestedHorizontalCollectionViewCell_)
  - NestedSingleStorySectionController (_StoryCell_)
- ContentSectionController (_NestedHorizontalCollectionViewCell_ + _SegmentedControlHeaderView_)
  - PhotosSectionController (_NestedVerticalCollectionViewCell_)
    - PhotoItemSectionController (_PhotoItemCell_)
  - ReelsSectionController (_NestedVerticalCollectionViewCell_)
    - ReelItemSectionController (_ReelItemCell_)
  - TaggedPhotosSectionController (_NestedVerticalCollectionViewCell_)
    - PhotoItemSectionController (_PhotoItemCell_)

Two things stand out to me:

1. When you swipe up from the top of the the screen, the deceleration of the top-level collection view is passed to the bottom-level collection view (_or scrolling is disabled on the bottom-level vertical CV_). 
2. More Importantly, getting the sizing down of the nested collection views. It seems when you swipe horizontally the contentOffset of the bottom-level vertical collection view (photos, reels, tagged) is set to 0, and the size is calculated from bottom to top before the entire `ContentSectionController`  is reloaded with the items size?

I have a very, very similar overall architecture for a UI I'm experimenting with and I can't seem to get the second item perfectly right.
